sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,379 @@
+device image and the function won't exported from this device image, while the
+other will be stored in a special device image for other `SYCL_EXTERNAL`
+functions and will be marked as exported there. Such duplication is need for
+two reasons:
+- We aim to make device images with kernels self-contained so no JIT linker
+  invocations would be needed if we have definitions of all called functions.
+  Also note that if AOT is requested, it would be impossible to link anything
+  at runtime.
+- We could export `SYCL_EXTERNAL` functions from device images with kernels,
+  but it would mean that when user's app calls `SYCL_EXTERNAL` function, it has
+  to link a whole kernel and all its dependencies - not only it increases the
+  amount of unnecessary linked code, but might also lead to build errors if the
+  kernel uses some features, which are not supported by target device (and they
+  are not used in the `SYCL_EXTERNAL` function).
+Besides separating `SYCL_EXTERNAL` functions from kernels, `sycl-post-link`
+can also distribute those functions into separate device images if device code
+split is requested. This is done by grouping them using `module-id` attribute.
+Non-`SYCL_EXTERNAL` functions used by `SYCL_EXTERNAL` functions with different
+`sycl-module-id` attributes are copied to device images corresponding to those
+`SYCL_EXTERNAL` functions to make them self-contained.
+In case one `SYCL_EXTERNAL` function uses another `SYCL_EXTERNAL` function
+with different `sycl-module-id` attribute, the second one is not copied to the
+device image with the first function, but dependency between those device images
+is recorder instead.
+
+After `SYCL_EXTERNAL` functions are arranged into a separate device image(s),
+all non-`SYCL_EXTERNAL` functions and `SYCL_EXTERNAL` functions left in device
+images with kernels marked with internal linkage to avoid multiple definition
+errors during runtime linking.
+Device images with `SYCL_EXTERNAL` functions will also get a list of names
+of exported functions attached to them through device image properties
+(described below).
+
+**NOTE**: If device code split is enabled, it seems reasonable to perform
+exports arrangement before device code split procedure.
+
+In order to collect information about imported symbols `sycl-post-link` looks
+through LLVM IR and for each declared but not defined symbol records its name,
+except the following cases:
+- Declarations with `__` prefix in demangled name are not recorded as imported
+  functions
+  - Declarations with `__spirv_*` prefix should not be recorded as dependencies
+  since they represent SPIR-V operations and will be transformed to SPIR-V
+  instructions during LLVM->SPIR-V translation.
+- Based on some attributes (which could be defined later) we may want to avoid
+  listing some functions as imported ones
+  - This is needed to have possibility to call device-specific builtins not
+    starting with `__` by forward-declaring them in DPC++ code
```

Comment:
Is there a danger that we will lose some error reporting?  Consider an application where a kernel calls a `SYCL_EXTERNAL` function that is not defined anywhere.  Currently, this causes an error from the static linker.  Will that still be the case once we implement this support for shared libraries?

It seems like this design might prevent the error from being reported when the static (ahead of time) linker runs.   According to the description above, a kernel reference to an undefined function will be assumed to be an import from a shared library, right?  I assume that means you will **not** get an error from the static linker, and you will instead get a runtime exception when the runtime fails to find an exported definition of the function.

Is my understanding correct?

Reply:
Thanks for the clarification.  I didn't realize that we depended on the host linker for these errors, but I see that you are correct.  (I verified this also with a test case.)  I agree that we will not lose any error reporting with this design.